### PR TITLE
Corrected result.data implementation.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
When looking into #4774 auto type promotion to objects raised a ValueError and was not tested. This PR fixes this and introduces the intended behavior as well as boosting performance slightly when qubit counts are < 64.

### CODE REVIEW

File: cirq-core/cirq/study/result_test.py [hunk 1]
```diff
@@ -166,6 +168,21 @@ def test_df():
...
+    result = cirq.ResultDict(
+        params=cirq.ParamResolver({}),
+        measurements={
+            'a': np.array([[0 for _ in range(76)]] * 10_000, dtype=bool),
+            'd': np.array([[1 for _ in range(76)]] * 10_000, dtype=bool),
+        },
+    )
+
+    assert np.all(result.data['a'] == 0)
+    assert np.all(result.data['d'] == 75557863725914323419135)
```
Comment (dstrain115) [inline cirq-core/cirq/study/result_test.py:181]: Can we add a comment or something to explain this magic number?

File: cirq-core/cirq/study/result.py [hunk 2]
```diff
@@ -142,14 +142,16 @@ def dataframe_from_measurements(measurements: Mapping[str, np.ndarray]) -> pd.Da
...
         # repetitions and a big endian integer for individual measurements.
-        converted_dict = {
-            key: [value.big_endian_bits_to_int(m_vals) for m_vals in val]
-            for key, val in measurements.items()
-        }
-        # Note that when a numpy array is produced from this data frame,
-        # Pandas will try to use np.int64 as dtype, but will upgrade to
-        # object if any value is too large to fit.
-        return pd.DataFrame(converted_dict, dtype=np.int64)
+        converted_dict = dict()
```
Comment (maffoo) [inline cirq-core/cirq/study/result.py:145]: nit: use dict literal
```suggestion
        converted_dict = {}
```

File: cirq-core/cirq/study/result.py [hunk 3]
```diff
@@ -142,14 +142,16 @@ def dataframe_from_measurements(measurements: Mapping[str, np.ndarray]) -> pd.Da
...
+        for key, bitstrings in measurements.items():
+            _, n = bitstrings.shape
+            i_type = object if n > 63 else np.int64
+            basis = 2 ** np.arange(bitstrings.shape[1], dtype=i_type)[::-1]
```
Comment (maffoo) [inline cirq-core/cirq/study/result.py:149]:
```suggestion
            basis = 2 ** np.arange(n, dtype=i_type)[::-1]
```

File: cirq-core/cirq/study/result.py [hunk 4]
```diff
@@ -142,14 +142,16 @@ def dataframe_from_measurements(measurements: Mapping[str, np.ndarray]) -> pd.Da
...
-            for key, val in measurements.items()
-        }
-        # Note that when a numpy array is produced from this data frame,
-        # Pandas will try to use np.int64 as dtype, but will upgrade to
-        # object if any value is too large to fit.
-        return pd.DataFrame(converted_dict, dtype=np.int64)
+        converted_dict = dict()
+        for key, bitstrings in measurements.items():
+            _, n = bitstrings.shape
+            i_type = object if n > 63 else np.int64
```
Comment (maffoo) [inline cirq-core/cirq/study/result.py:148]: nit: call this `dtype` instead
```suggestion
            dtype = object if n > 63 else np.int64
```

File: cirq-core/cirq/study/result.py [hunk 5]
```diff
@@ -142,14 +142,16 @@ def dataframe_from_measurements(measurements: Mapping[str, np.ndarray]) -> pd.Da
...
+            basis = 2 ** np.arange(bitstrings.shape[1], dtype=i_type)[::-1]
+            converted_dict[key] = np.sum(basis * bitstrings, axis=1)
+
+        # Use objects to accomodate more than 64 qubits if needed.
+        ret_type = object if any(bs.shape[1] > 63 for _, bs in measurements.items()) else np.int64
```
Comment (maffoo) [inline cirq-core/cirq/study/result.py:153]: nit: call this `dtype` instead
```suggestion
        dtype = object if any(bs.shape[1] > 63 for _, bs in measurements.items()) else np.int64
```

File: cirq-core/cirq/study/result_test.py (hunk 1 above)
Comment (maffoo) [reply cirq-core/cirq/study/result_test.py:181]: Would perhaps be clearer if written in hex (underscores are optional, just for grouping):
```suggestion
    assert np.all(result.data['d'] == 0xfff_ffffffff_ffffffff)
```

**Review by maffoo — APPROVED** [review]
Minor comments, then LGTM

File: cirq-core/cirq/study/result_test.py (hunk 1 above)
Comment (mpharrigan) [reply cirq-core/cirq/study/result_test.py]: ah you stole my suggestion

**Pull request merged**